File: benchmark/go/stencil_2d.go
package main

import "fmt"

func idx(x int, y int, width int) int {
	return y*width + x
}

func main() {
	width := 256
	height := 256
	steps := 5
	total := width * height

	current := make([]int64, total)
	next := make([]int64, total)
	for i := 0; i < total; i++ {
		current[i] = ((int64(i)*48271 + 97) % 131071) - 65535
	}

	for step := 0; step < steps; step++ {
		for y := 1; y < height-1; y++ {
			for x := 1; x < width-1; x++ {
				center := current[idx(x, y, width)]
				north := current[idx(x, y-1, width)]
				south := current[idx(x, y+1, width)]
				west := current[idx(x-1, y, width)]
				east := current[idx(x+1, y, width)]
				next[idx(x, y, width)] = center*4 + north + south + west + east
			}
		}

		for i := 0; i < total; i++ {
			current[i] = next[i] % 104729
		}
	}

	checksum := int64(0)
	for i, value := range current {
		checksum += value * ((int64(i) % 31) + 1)
	}

	fmt.Println(checksum)
}

File: benchmark/rust/stencil_2d.rs
#[inline(always)]
fn idx(x: usize, y: usize, width: usize) -> usize {
    y * width + x
}

fn main() {
    let width = 256_usize;
    let height = 256_usize;
    let steps = 5_usize;
    let total = width * height;

    let mut current = Vec::with_capacity(total);
    let mut next = vec![0_i64; total];
    for i in 0..total {
        current.push((((i as i64) * 48_271 + 97) % 131_071) - 65_535);
    }

    for _ in 0..steps {
        for y in 1..(height - 1) {
            for x in 1..(width - 1) {
                let center = current[idx(x, y, width)];
                let north = current[idx(x, y - 1, width)];
                let south = current[idx(x, y + 1, width)];
                let west = current[idx(x - 1, y, width)];
                let east = current[idx(x + 1, y, width)];
                next[idx(x, y, width)] = center * 4 + north + south + west + east;
            }
        }

        for i in 0..total {
            current[i] = next[i] % 104_729;
        }
    }

    let mut checksum = 0_i64;
    for (i, value) in current.iter().enumerate() {
        checksum += *value * (((i as i64) % 31) + 1);
    }

    println!("{checksum}");
}
